File: projects/github-tracker/src/github_tracker.py
import os
import json
import urllib.request
from datetime import datetime, timedelta
from dotenv import load_dotenv

load_dotenv()

USERNAME = os.getenv('GITHUB_USERNAME')
TELEGRAM_TOKEN = os.getenv('TELEGRAM_TOKEN')
CHAT_ID = os.getenv('TELEGRAM_CHAT_ID')

def get_github_activity():
    url = f'https://api.github.com/users/{USERNAME}/events/public'
    req = urllib.request.Request(url, headers={'User-Agent': 'Python'})
    response = urllib.request.urlopen(req)
    events = json.loads(response.read())

    one_day_ago = datetime.now() - timedelta(hours=24)
    recent_pushes = [
        event for event in events
        if event['type'] == 'PushEvent'
        and datetime.strptime(event['created_at'], '%Y-%m-%dT%H:%M:%SZ') > one_day_ago
    ]
    return recent_pushes

def create_summary(pushes):
    if len(pushes) == 0:
        return 'No GitHub activity in the last 24 hours.'

    total_pushes = 0
    repos = set()

    for push in pushes:
        total_pushes += 1
        repos.add(push['repo']['name'])

    repo_list = ', '.join(repos)
    return f'GitHub Activity Summary:\nTotal pushes: {total_pushes}\nRepos: {repo_list}'

def save_to_file(summary):
    date = datetime.now().strftime('%Y-%m-%d')
    entry = f'\n--- {date} ---\n{summary}\n'
    with open('github-activity-log.txt', 'a') as f:
        f.write(entry)

def send_telegram(summary):
    url = f'https://api.telegram.org/bot{TELEGRAM_TOKEN}/sendMessage'
    data = json.dumps({'chat_id': CHAT_ID, 'text': summary}).encode('utf-8')
    req = urllib.request.Request(url, data=data, headers={'Content-Type': 'application/json'})
    urllib.request.urlopen(req)

print('Checking GitHub activity...')
pushes = get_github_activity()
summary = create_summary(pushes)
save_to_file(summary)
print('Saved to file!')
send_telegram(summary)
print('Sent to Telegram!')
print(summary)

File: projects/github-tracker/src/github_tracker.js
 require('dotenv').config();
  const fs = require('fs');

  const USERNAME = process.env.GITHUB_USERNAME;
  const TELEGRAM_TOKEN = process.env.TELEGRAM_TOKEN;
  const CHAT_ID = process.env.TELEGRAM_CHAT_ID;

  // Fetch recent GitHub activity
  async function getGitHubActivity() {
    const response = await fetch(`https://api.github.com/users/${USERNAME}/events/public`);
    const events = await response.json();

    // Filter only push events (commits) from last 24 hours
    const oneDayAgo = new Date(Date.now() - 24 * 60 * 60 * 1000);
    const recentPushes = events.filter(event =>
      event.type === 'PushEvent' && new Date(event.created_at) > oneDayAgo
    );

    return recentPushes;
  }

  // Create a summary from the activity
  function createSummary(pushes) {
    if (pushes.length === 0) {
      return 'No GitHub activity in the last 24 hours.';
    }

    let totalCommits = 0;
    const repos = new Set();

    pushes.forEach(push => {
      totalCommits += push.payload.commits.length;
      repos.add(push.repo.name);
    });

    const repoList = Array.from(repos).join(', ');
    return `GitHub Activity Summary:\nTotal commits: ${totalCommits}\nRepos: ${repoList}`;
  }

  // Save summary to a file
  function saveToFile(summary) {
    const date = new Date().toLocaleDateString();
    const entry = `\n--- ${date} ---\n${summary}\n`;
    fs.appendFileSync('github-activity-log.txt', entry);
  }

  // Send summary to Telegram
  async function sendTelegram(summary) {
    const url = `https://api.telegram.org/bot${TELEGRAM_TOKEN}/sendMessage`;
    await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ chat_id: CHAT_ID, text: summary })
    });
  }

  // Run everything
  async function run() {
    console.log('Checking GitHub activity...');
    const pushes = await getGitHubActivity();
    const summary = createSummary(pushes);
    saveToFile(summary);
    console.log('Saved to file!');
    await sendTelegram(summary);
    console.log('Sent to Telegram!');
    console.log(summary);
  }

  run();